Propagate dynamic int on `__setitem__`.

ysiraichi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106208
* __->__ #105923

Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc @ezyang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

ysiraichi committed:
Propagate dynamic int on `__setitem__`.

Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

[ghstack-poisoned]

ysiraichi committed:
Rebased. on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

albanD commented on torch/csrc/autograd/python_variable_indexing.cpp:
```diff
@@ -468,12 +469,16 @@ int THPVariable_setitem(PyObject* self, PyObject* index, PyObject* py_value) {
-    dispatch_set_item(
-        self_, {at::indexing::TensorIndex(THPUtils_unpackLong(index))}, value);
+    auto symint = torch::is_symint(index)
+        ? SymInt(static_cast<SymNode>(
```
There no simpler way to extract the SymInt from the PyObject ??

Skylion007 replied:
Yeah do we not have a working cast method implementation that does? @ezyang ?

ezyang replied:
py::cast should work E.g. `return py::cast<c10::SymInt>(index);`

ysiraichi committed:
Fix compilation error. on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented on aten/src/ATen/TensorIndexing.h:
```diff
@@ -230,22 +231,24 @@ static inline Tensor applySlice(
     auto size = (*self_sizes)[dim];
     TORCH_CHECK_INDEX(
-        size >= -index && size > index,
+        size >= -index_value && size > index_value,
```
can't you just put some guards in here? Seems reasonable to do

ysiraichi replied:
Or actually issuing it to `ShapeEnv` with `index < size`?

ezyang replied:
Issue it to the shape env

ysiraichi committed:
Update on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented:
needs test

ysiraichi committed:
Update on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi committed:
Update on "Propagate dynamic int on `__setitem__`."


Fix: #105533

This PR propagates dynamic ints used as indices for `__setitem__`. In summary, we:

- Replace the integer type for `TensorIndex` (both the enum and the corresponding
functions)
- Accordingly modify _python_variable_indexing.cpp_

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ysiraichi commented:
@pytorchbot merge -i

lezcano commented:
@pytorchbot merge -i
